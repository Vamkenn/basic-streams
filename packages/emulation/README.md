# [@basic-streams](https://github.com/rpominov/basic-streams)/emulation

Although the version is above `1.0` this package is not stable. Please use with
caution API may change.
