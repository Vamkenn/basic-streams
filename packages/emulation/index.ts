import {Stream} from "@basic-streams/stream"

export class Value<T> {
  constructor(public readonly value: T) {}
}

export class TimeSpan {
  constructor(public readonly ms: number) {}
}

export type Timeline<T> = Array<TimeSpan | Value<T>>

export class Event<T> {
  constructor(
    public readonly time: number,
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }
}

export class EventsList<T> {
  constructor(public readonly items: Array<Event<T>>) {}

  static fromTimeline<T>(items: Timeline<T>, currentTime = 0): EventsList<T> {
    const events: Array<Event<T>> = []
    for (const item of items) {
      if (item instanceof TimeSpan) {
        currentTime = currentTime + item.ms
      } else {
        events.push(new Event(currentTime, item.value))
      }
    }
    return new EventsList(events)
  }

  withCb(cb: (value: T) => void): EventsList<T> {
    return new EventsList(
      this.items.map(item => new Event(item.time, item.value, cb)),
    )
  }

  merge(another: EventsList<T>): EventsList<T> {
    const events: Array<Event<T>> = []
    let indexA = 0
    let indexB = 0
    const itemsA = this.items
    const itemsB = another.items
    while (itemsA.length > indexA || itemsB.length > indexB) {
      if (
        itemsA.length !== indexA &&
        (itemsB.length === indexB || itemsA[indexA].time <= itemsB[indexB].time)
      ) {
        events.push(itemsA[indexA])
        indexA++
      } else {
        events.push(itemsB[indexB])
        indexB++
      }
    }
    return new EventsList(events)
  }

  takeOne(): {event: Event<T> | void; rest: EventsList<T>} {
    const [event, ...rest] = this.items
    return {event, rest: new EventsList(rest)}
  }

  toJSON() {
    return this.items.map(item => ({time: item.time, value: item.value}))
  }

  static jestSerializer = {
    test(x: any) {
      return x instanceof EventsList
    },

    // for outdated TS typings
    print(x: any): string {
      return ""
    },

    serialize<Refs, Config extends {indent: string}, T>(
      val: EventsList<T>,
      config: Config,
      indentation: string,
      depth: number,
      refs: Refs,
      printer: (
        x: T,
        config: Config,
        indentation: string,
        depth: number,
        refs: Refs,
      ) => string,
    ) {
      const separator = "\n"

      function printItem(item: Event<T>) {
        const indentation1 = indentation + config.indent
        const value = printer(item.value, config, indentation1, depth + 1, refs)
        return indentation1 + item.time + ": " + value
      }

      return `EventsList(${
        val.items.length === 0
          ? ""
          : separator + val.items.map(printItem).join(separator) + separator
      })`
    },
  }
}

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(
  generator: (
    createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
  ) => Stream<T>,
  maxTime = Infinity,
): EventsList<T> {
  const state = {
    time: 0,
    toProduce: new EventsList([]) as EventsList<any>,
    result: [] as Array<Event<T>>,
  }
  const resultStream = generator((...timeline) => {
    return cb => {
      // wrapping cb to make it a unique value
      // that we will compare to in unsuscribe
      const _cb = (x: any) => cb(x)
      state.toProduce = state.toProduce.merge(
        EventsList.fromTimeline(timeline, state.time).withCb(_cb),
      )
      return () => {
        state.toProduce = new EventsList(
          state.toProduce.items.filter(item => item.cb !== _cb),
        )
      }
    }
  })
  resultStream(value => {
    state.result.push(new Event(state.time, value))
  })
  while (true) {
    const {event, rest} = state.toProduce.takeOne()
    if (state.time >= maxTime || !event) {
      return new EventsList(state.result)
    }
    state.time = event.time
    state.toProduce = rest
    event.callCb()
  }
}

export function laterMock(
  createStream: <U>(...timeline: Timeline<U>) => Stream<U>,
) {
  return function later<T>(time: number, value?: T): Stream<T> {
    return createStream(t(time), v(value as any))
  }
}
