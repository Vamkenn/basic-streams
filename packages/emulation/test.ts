import {EventsList, emulate, t, v} from "./index"

expect.addSnapshotSerializer(EventsList.jestSerializer)

describe("emulate", () => {
  test("map", () => {
    expect(
      emulate(create => {
        return cb => create(t(10), v(1))(x => cb(x + 1))
      }),
    ).toMatchSnapshot()
  })

  test("filter", () => {
    expect(
      emulate(create => {
        return cb =>
          create(v(1), t(10), v(0), t(10), v(1))(x => {
            if (x > 0) cb(x)
          })
      }),
    ).toMatchSnapshot()
  })

  test("order of simultaneous values", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1))
        const b = create(t(10), v(2))
        return cb => {
          const disposeA = a(cb)
          const disposeB = b(cb)
          return () => {
            disposeA()
            disposeB()
          }
        }
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(15), v(null))
        return cb => b(a(cb))
      }),
    ).toMatchSnapshot()
  })

  test("unsubscribe doesn't remove other values with same cb", () => {
    expect(
      emulate(create => {
        const a = create(t(10), v(1), t(10), v(0))
        const b = create(t(10), v(1), t(10), v(0))
        const unsubscriber = create(t(15), v(null))
        return cb => {
          a(cb)
          unsubscriber(b(cb))
          return () => {}
        }
      }),
    ).toMatchSnapshot()
  })
})
