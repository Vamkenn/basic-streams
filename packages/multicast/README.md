# @basic-streams/multicast

TODO: fill up README
