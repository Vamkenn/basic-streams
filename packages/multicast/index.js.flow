// @flow

import type {Stream} from "@basic-streams/stream"

declare export default function multicast<T>(stream: Stream<T>): Stream<T>
