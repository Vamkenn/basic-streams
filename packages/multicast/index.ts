import {Stream} from "@basic-streams/stream"

export default function multicast<T>(stream: Stream<T>): Stream<T> {
  let cbs: ((x: T) => void)[] = []
  const push = (x: T) => {
    cbs.forEach(sink => {
      if (cbs.indexOf(sink) !== -1) {
        sink(x)
      }
    })
  }
  let unsub: null | (() => void) = null
  return cb => {
    let disposed = false
    cbs = [...cbs, cb]
    if (cbs.length === 1) {
      unsub = stream(push)
    }
    return () => {
      if (disposed) {
        return
      }
      disposed = true
      const index = cbs.indexOf(cb)
      cbs = [...cbs.slice(0, index), ...cbs.slice(index + 1, cbs.length)]
      if (cbs.length === 0 && unsub !== null) {
        unsub()
        unsub = null
      }
    }
  }
}
