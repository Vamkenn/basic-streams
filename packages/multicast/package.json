{
  "name": "@basic-streams/multicast",
  "version": "0.0.12",
  "description": "multicast operator for basic-streams",
  "keywords": [
    "basic-streams",
    "multicast"
  ],
  "author": "Roman Pominov <rpominov@gmail.com>",
  "main": "./index.js",
  "types": "./index.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rpominov/basic-streams.git"
  },
  "bugs": {
    "url": "https://github.com/rpominov/basic-streams/issues"
  },
  "homepage": "https://github.com/rpominov/basic-streams/packages/multicast",
  "dependencies": {
    "@basic-streams/stream": "^0.0.8"
  }
}
