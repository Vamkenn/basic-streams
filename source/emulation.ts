import {Stream} from "./index"

export class Value<T> {
  constructor(
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }

  toJSON() {
    return this.value
  }
}

export class TimeSpan {
  constructor(public readonly ms: number) {}

  toJSON() {
    return `after ${this.ms} ms...`
  }
}

type TimelineItem<T> = TimeSpan | Value<T>

export class Timeline<T> {
  constructor(
    public readonly items: ReadonlyArray<TimelineItem<T>>,
    dontCompact?: boolean,
  ) {
    if (dontCompact) {
      return this
    }

    const finalItems: Array<TimelineItem<T>> = []
    for (const item of this.items) {
      if (item instanceof TimeSpan) {
        if (item.ms === 0) {
          continue
        }
        if (finalItems.length > 0) {
          const last = finalItems[finalItems.length - 1]
          if (last instanceof TimeSpan) {
            finalItems[finalItems.length - 1] = new TimeSpan(last.ms + item.ms)
            continue
          }
        }
      }
      finalItems.push(item)
    }

    if (
      finalItems.length > 0 &&
      finalItems[finalItems.length - 1] instanceof TimeSpan
    ) {
      finalItems.pop()
    }

    return new Timeline(finalItems, true)
  }

  withCb(cb: (value: T) => void): Timeline<T> {
    return new Timeline(
      this.items.map(
        item => (item instanceof TimeSpan ? item : new Value(item.value, cb)),
      ),
    )
  }

  merge<U>(
    another: Timeline<U>,
    result: Timeline<T | U> = new Timeline([]),
  ): Timeline<T | U> {
    const a = this
    const b = another
    if (a.items.length === 0 && b.items.length === 0) {
      return result
    }
    if (a.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(b.items as Array<TimelineItem<T | U>>),
      ])
    }
    if (b.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(a.items as Array<TimelineItem<T | U>>),
      ])
    }
    const timeToValueA = a.timeToValue()
    const timeToValueB = b.timeToValue()
    if (timeToValueA === Infinity && timeToValueB === Infinity) {
      return new Timeline([
        new TimeSpan(Math.max(a.timeToEnd(), b.timeToEnd())),
        ...result.items,
      ])
    }
    const timeToValue = Math.min(timeToValueA, timeToValueB)
    const subtractA = a.subtractTime(timeToValue)
    const subtractB = b.subtractTime(timeToValue)
    return subtractA.rest.merge(
      subtractB.rest,
      new Timeline([
        ...result.items,
        new TimeSpan(timeToValue),
        ...(subtractA.values as Array<Value<T | U>>),
        ...(subtractB.values as Array<Value<T | U>>),
      ]),
    )
  }

  isEmpty(): boolean {
    return this.items.length === 0
  }

  takeOne(): {item: TimelineItem<T>; rest: Timeline<T>} {
    if (this.isEmpty()) {
      throw new Error("cannot takeOne() from an empty schedule")
    }
    const [top, ...rest] = this.items
    return {item: top, rest: new Timeline(rest)}
  }

  subtractTime(ms: number): {values: Value<T>[]; rest: Timeline<T>} {
    if (this.isEmpty()) {
      return {rest: this, values: []}
    }
    const {item, rest} = this.takeOne()
    if (item instanceof Value) {
      const r = rest.subtractTime(ms)
      return {rest: r.rest, values: [item, ...r.values]}
    }
    return item.ms <= ms
      ? rest.subtractTime(ms - item.ms)
      : {
          rest: new Timeline([new TimeSpan(item.ms - ms), ...rest.items]),
          values: [],
        }
  }

  timeToEnd(): number {
    return this.items.reduce(
      (result, x) => result + (x instanceof TimeSpan ? x.ms : 0),
      0,
    )
  }

  timeToValue(): number {
    if (this.isEmpty()) {
      return Infinity
    }
    const {item, rest} = this.takeOne()
    return item instanceof Value ? 0 : item.ms + rest.timeToValue()
  }

  toJSON() {
    return this.items
  }
}

type EmulationGenerator<T> = (
  createStream: <U>(...scheduleItems: Array<TimelineItem<U>>) => Stream<U>,
) => Stream<T>

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(generator: EmulationGenerator<T>): Timeline<T> {
  const state = {
    schedule: new Timeline([]),
    result: [] as Array<TimelineItem<T>>,
  }

  const resultStream = generator((...scheduleItems) => {
    return cb => {
      state.schedule = state.schedule.merge(
        new Timeline(scheduleItems).withCb(cb),
      )
      return () => {
        state.schedule = new Timeline(
          state.schedule.items.filter(
            item => item instanceof TimeSpan || item.cb !== cb,
          ),
        )
      }
    }
  })

  resultStream(value => state.result.push(new Value(value)))

  while (state.schedule.timeToValue() !== Infinity) {
    const time = state.schedule.timeToValue()
    const {rest, values} = state.schedule.subtractTime(time)
    state.schedule = rest
    state.result.push(new TimeSpan(time))
    values.forEach(item => item.callCb())
  }

  return new Timeline(state.result)
}
