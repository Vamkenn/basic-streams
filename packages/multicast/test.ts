import {EventsList, emulate, t, v} from "../emulation"
import multicast from "./index"
import ofMany from "../of-many"

expect.addSnapshotSerializer(EventsList.jestSerializer)

function noop() {}

test("first subscribers gets sync events", () => {
  const cb = jest.fn()
  multicast(ofMany([1, 2]))(cb)
  expect(cb.mock.calls).toMatchSnapshot()
})

test("second subscriber doesn't get sync events", () => {
  const cb = jest.fn()
  const stream = multicast(ofMany([1, 2]))
  stream(noop)
  stream(cb)
  expect(cb.mock.calls).toMatchSnapshot()
})

test("first and second sunscribers get same async events", () => {
  const cb1 = jest.fn()
  const cb2 = jest.fn()
  let inCb = null
  const stream = multicast(_inCb => {
    inCb = _inCb
    return () => {
      inCb = null
    }
  })
  stream(cb1)
  stream(cb2)
  inCb(1)
  inCb(2)
  expect(cb1.mock.calls).toMatchSnapshot()
  expect(cb2.mock.calls).toMatchSnapshot()
})

test("subscriber doesn't get event after unsub() in response to that event", () => {
  const cb = jest.fn()
  let inCb = null
  let stream = multicast(_inCb => {
    inCb = _inCb
    return () => {
      inCb = null
    }
  })
  let unsub = null
  stream(x => {
    if (x === 2) {
      unsub && unsub()
    }
  })
  unsub = stream(cb)
  inCb(1)
  inCb(2)
  inCb(3)
  expect(cb.mock.calls).toMatchSnapshot()
})

test("preserves disposer (single subscriber)", () => {
  const disposer = jest.fn()
  multicast(cb => disposer)(noop)()
  expect(disposer.mock.calls).toMatchSnapshot()
})

test("preserves disposer (two subscribers)", () => {
  const disposer = jest.fn()
  const stream = multicast(cb => disposer)
  const unsub1 = stream(noop)
  const unsub2 = stream(noop)
  unsub1()
  unsub2()
  expect(disposer.mock.calls).toMatchSnapshot()
})

test("unsub doesn't remove another version of same subscriber", () => {
  const subscriber = jest.fn()
  let inCb = null
  const stream = multicast(_inCb => {
    inCb = _inCb
    return () => {
      inCb = null
    }
  })
  const unsub = stream(subscriber)
  stream(subscriber)
  unsub()
  unsub() // should be noop
  inCb(1)
  expect(subscriber.mock.calls).toMatchSnapshot()
})
