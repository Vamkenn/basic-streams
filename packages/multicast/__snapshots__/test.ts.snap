// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first and second sunscribers get same async events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`first and second sunscribers get same async events 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`first subscribers gets sync events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`preserves disposer (single subscriber) 1`] = `
Array [
  Array [],
]
`;

exports[`preserves disposer (two subscribers) 1`] = `
Array [
  Array [],
]
`;

exports[`second subscriber doesn't get sync events 1`] = `Array []`;

exports[`subscriber doesn't get event after unsub() in response to that event 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`unsub doesn't remove another version of same subscriber 1`] = `
Array [
  Array [
    1,
  ],
]
`;
